components/cdc/src/endpoint.rs

```diff
@@ -350,14 +349,103 @@ impl ResolvedRegionHeap {
     fn is_empty(&self) -> bool {
         self.heap.is_empty()
     }
+}
 
-    fn clear(&mut self) {
-        self.heap.clear();
-    }
+#[derive(Default, Debug)]
+pub(crate) struct Advance {
+    // multiplexing means one region can be subscribed multiple times in one `Conn`,
+    // in which case progresses are grouped by (ConnId, request_id).
+    pub(crate) multiplexing: HashMap<(ConnId, u64), ResolvedRegionHeap>,
+
+    // exclusive means one region can only be subscribed one time in one `Conn`,
+    // in which case progresses are grouped by ConnId.
+    pub(crate) exclusive: HashMap<ConnId, ResolvedRegionHeap>,
+
+    // To be compatible with old TiCDC client before v4.0.8.
+    // TODO(qupeng): we can deprecate support for too old TiCDC clients.
+    // map[(ConnId, region_id)]->request_id.
+    pub(crate) dispersed: HashMap<(ConnId, u64), u64>,
+
+    pub(crate) scan_finished: usize,
+
+    pub(crate) blocked_on_scan: usize,
+
+    pub(crate) blocked_on_locks: usize,
+}
+
+impl Advance {
+    fn emit_resolved_ts(&mut self, connections: &HashMap<ConnId, Conn>) -> (u64, TimeStamp) {
+        let handle_send_result = |conn: &Conn, res: std::result::Result<(), SendError>| -> bool {
+            match res {
+                Ok(_) => return true,
+                Err(SendError::Disconnected) => {
+                    debug!("cdc send event failed, disconnected";
```

Comment:
will this output too much logs? It looks that `disconnected` may happens more frequently than `full` and `congested`, since the latter two log level is set to info, but this one set to debug.

Reply:
By default `info` will be called 1 time in 1 second for every connection. I think it's OK. We need this log to debug.